### PR TITLE
vm-monitor: Switch from memory.high to polling memory.stat

### LINKED ISSUE #5444: Epic: Fix vm-monitor cgroup memory.high throttling
Marked as epic because this has been ongoing for a little while, and I expect it will take another ~1.5 weeks to completely resolve, due to size of the initial PR, and time to review follow-ups.

## Technical background

Neon's autoscaling feature relies on a handful of components managed by the autoscaling team. One of these is the `vm-monitor` (defined in [`libs/vm_monitor`](../tree/main/libs/vm_monitor)) and runs _inside_ each VM. It can be provided either as a binary (for the autoscaling repo's CI) or embedded into an existing tokio runtime (like it is with compute_ctl).

One way we provide better guarantees around the speed of upscaling is by running postgres in its own cgroup and listening for "memory.high events" on that cgroup. This allows the vm-monitor to be notified ~immediately when a memory threshold is exceeded, so we can make timely upscaling requests without excessively polling.

These memory.high events are represented as changes to `high` field of the cgroup's `memory.events` file, which generates the appropriate modification events, etc. In order to generate these events, the cgroup must have the `memory.high` value set to _some_ number of bytes (and then, exceeding that value will generate events, etc.)

Beyond generating events, the _primary_ purpose of memory.high is actually to provide a threshold above which the kernel should start reclaiming memory from processes in the cgroup _and_ throttling all processes in the cgroup.

It turns out that this throttling is actually _independent_ of the memory reclamation and, if left unchecked for more than a couple seconds, quickly becomes quite severe (on the order of 1000x slower).

## Historical context

At time of writing, the vm-monitor is actually a relatively recent creation, first used via neondatabase/autoscaling#362 (2023-07-03), and finally taking over from its predecessor with neondatabase/autoscaling#442 (2023-07-17).

Before the vm-monitor, we had the vm-informant (originally introduced 2023-01-10, with neondatabase/autoscaling#8), which performed the exact same role, but was written in Go, and developed inside the autoscaling repo.

The vm-informant had _implemented_ a similar cgroup event-based upscaling mechanism, and we had lightly tested it, but actually _enabling_ it required a small fix to how VMs were created by the control plane, and e2e tests on those changes always had opaque failures for reasons that were unclear at the time (see neondatabase/cloud#4557 and https://github.com/neondatabase/cloud/pull/5143#issuecomment-1575689390).

Nevertheless, the vm-informant's original implementation was found to _heavily_ throttle the postgres cgroup if the VM wasn't upscaled (so severely that we originally thought the cgroup was frozen). This was due exclusively to the throttling by the kernel from memory.high itself, so we changed the vm-informant to gradually increase memory.high when we hit it, to avoid this throttling, in neondatabase/autoscaling#223 (2023-05-15). See also: the [original issue](https://github.com/neondatabase/autoscaling/issues/44).

The original implementation of the vm-monitor did not have this "gradually increase memory.high" logic, and after we finally _first_ enabled the cgroup integration using the vm-monitor (#4920), we started seeing VMs stalling due to the same memory.high throttling that previously existed with the vm-informant:

- [2023-09-09](https://neondb.slack.com/archives/C03H1K0PGKH/p1694287808046179?thread_ts=1694271790.580099&cid=C03H1K0PGKH)
- [2023-09-10](https://neondb.slack.com/archives/C03F5SM1N02/p1694379934417009)
- [2023-09-12](https://neondb.slack.com/archives/C03H1K0PGKH/p1694511932560659)

Typical symptoms were that it took ~20s to connect to postgres with psql _from inside the VM_, and `\d+` took longer than we cared to wait for (> 2 minutes).

We initially fixed this by reimplementing the vm-informant's updated logic (see #5303; every time we get a memory.high event, increase memory.high by a small amount), but it turned out this wasn't a complete fix, due to some remaining quirks - some fixed, and some work still required (more detail below).

- [2023-09-20](https://neondb.slack.com/archives/C03F5SM1N02/p1695218132208249)
- [2023-09-21](https://neondb.slack.com/archives/C03F5SM1N02/p1695314613696659)
- [2023-09-22](https://neondb.slack.com/archives/C05L7D1JAUS/p1695408776416279)
- [2023-09-24](https://neondb.slack.com/archives/C03H1K0PGKH/p1695579549367099?thread_ts=1695579487.966629&cid=C03H1K0PGKH)
- [2023-10-04](https://neondb.slack.com/archives/C03F5SM1N02/p1696383991691429)

**This issue exists to track the rest of the work required to fix any remaining instances of this throttling.**

## Planned changes

At a high level, there's two pieces of work remaining, at time of writing (check the task list below). There's also a secret third piece that's blocking the other two.

First, we need to fix the bug that allows memory.high to be decreased when the vm-monitor is notified of increases to the VM's memory size. Ideally this _would_ be fairly straightforward, but there's some existing raciness between the logic handling upscaling/downscaling and the cgroup watcher's own "increase memory.high if we need to" that should be resolved first (or, alongside it) — in short, they both touch memory.high.

Second: We need to provide a mechanism outside downscaling to decrease the cgroup's memory.high value. Periodically decreasing memory.high if we can (assuming current memory usage is some threshold below target value) should suffice here. The cgroup event watcher is already responsible for increasing memory.high when we hit it to avoid throttling, and so it's kind of the ideal choice for managing this as well.

Therefore, before either of these, we need to move memory.high handling into the cgroup event watcher with the end result being that:

1. It's responsible for approving downscaling (or not); and
2. Responding to an upscaling message blocks on the cgroup event watcher executing the change

In addition to fixing the existing raciness, that should make the rest of this _much_ simpler to implement.

## Tasks

```[tasklist]
- [ ] https://github.com/neondatabase/neon/pull/5303
- [ ] https://github.com/neondatabase/neon/pull/5348
- [ ] https://github.com/neondatabase/neon/pull/5488
- [ ] bugfix: Don't decrease memory.high on upscale
- [ ] bugfix: Periodically try to decrease memory.high if it was previously artificially increased
```

### PR DESCRIPTION
**Pull request by sharnoff**
tl;dr it's really hard to avoid throttling from memory.high, and it counts tmpfs & page cache usage, so it's also hard to make sense of.

In the interest of fixing things quickly with something that should be *good enough*, this PR switches to instead periodically fetch memory statistics from the cgroup's memory.stat and use that data to determine if and when we should upscale.

This PR fixes #5444, which has a lot more detail on the difficulties we've hit with memory.high. This PR also supersedes #5488.

---

Opened this quickly to (hopefully) save on time spent waiting on review. ~~Still need to (unfortunately) do some manual tests that this actually works as intended. Also still need to self-review the code.~~

Currently this PR polls every 100ms and uses a rolling average of the last 500ms to make decisions, periodically logging the last 2s of values collected. This could be improved, and I'm open to suggestions.

### CODE REVIEW

**Comment by sharnoff** [comment]
Tested this on staging by manually deleting & recreating a VM with the image built by CI, using this script that IIRC I've shared elsewhere:

```
#!/usr/bin/env bash

set -eu -o pipefail

if [ "$#" -ne 2 ]; then
    echo "USAGE: $0 <VirtualMachine name> <image>"
    exit 1
fi

vm="$1"
image="$2"

vm_json="$(kubectl get neonvm "$vm" -o json | jq --arg img "$image" '.spec.guest.rootDisk.image = $img')"

# delete the VM and recreate it
kubectl delete neonvm "$vm"
echo "$vm_json" | kubectl apply -f -
sh
```

Results from testing: I was unable to get "normal" pgbench to trigger upscaling — and memory seemed to stay at normal amounts. When I ran the synthetic "high CPU usage" pgbench query, upscaling and downscaling performed as normal.

Logs were far too verbose. It's currently logging recent memory stats about once every 0.5s. It should be more like 2-5s, at least.

At this point, need to test on locally with the autoscaling repo's tooling for synthetic memory usage, but presuming that passes, then we should be good.

**Comment by problame** [comment]
> At this point, need to test on locally with the autoscaling repo's tooling for synthetic memory usage, but presuming that passes, then we should be good.

I assume theses tests happened?

**Comment by problame** [comment]
This needs updating:

https://github.com/neondatabase/neon/blob/85d6d9dc85005dddef2921feba24ea0946fd6fe4/libs/vm_monitor/README.md#L12-L13

File: libs/vm_monitor/src/runner.rs [hunk 1]
```diff
@@ -408,33 +410,53 @@ impl Runner {
...
                     }
                 }
-                // we need to propagate an upscale request
-                request = self.dispatcher.request_upscale_events.recv(), if self.cgroup.is_some() => {
-                    if request.is_none() {
-                        bail!("failed to listen for upscale event from cgroup")
+
+                // New memory stats from the cgroup, *may* need to request upscaling, if we've
+                // exceeded the threshold
+                result = self.cgroup.as_mut().unwrap().watcher.changed(), if self.cgroup.is_some() => {
```
Comment (problame) [inline libs/vm_monitor/src/runner.rs:416]: The docs say

> Additionally, each branch may include an optional if precondition. If the precondition returns false, then the branch is disabled. The provided <async expression> is still evaluated but the resulting future is never polled. This capability is useful when using select! within a loop.

https://docs.rs/tokio/latest/tokio/macro.select.html

So, I guess the `.unwrap()` here would panci?

(I haven't used this feature of select! before.)

File: libs/vm_monitor/src/runner.rs [hunk 2]
```diff
@@ -62,16 +66,40 @@ pub struct Config {
...
     }
 }
 
+impl Config {
+    fn cgroup_threshold(&self, total_mem: u64, file_cache_disk_size: u64) -> u64 {
+        // If the file cache is in memory, then it will count towards shmem usage of the cgroup,
+        // and thus be non-reclaimable, so we should allow for additional memory usage. Otherwise
+        // (i.e. if the file cache is on disk), the cgroup's non-reclaimable memory usage will not
+        // include the file cache, and so non-reclaimable memory usage from the cgroup *plus*
+        // memory reserved for the file cache should try to be less than total memory.
```
Comment (problame) [inline libs/vm_monitor/src/runner.rs:93]:
```suggestion
        // and thus be non-reclaimable, so we should allow for additional memory usage.
        //
        // If the file cache sits on disk, our desired stable system state is for it to
        // be fully page-cached (the file cache should only be paged to/from disk in situations
        // where we can't upscale fast enough).
        // Page-cached are reclaimable memory, so, we need to lower the threshold
        // for non-reclaimable memory so we scale up _before_ the kernel starts paging
        // out he file cache.
```

File: libs/vm_monitor/src/runner.rs [hunk 3]
```diff
@@ -62,16 +66,40 @@ pub struct Config {
...
+        let memory_remaining_for_cgroup = total_mem.saturating_sub(file_cache_disk_size);
+
+        // Guarantee at least some wiggle room, even if there is none from the file cache
+        let max_remaining = (total_mem as f64 * (1.0 - self.cgroup_min_overhead_fraction)) as u64;
```
Comment (problame) [inline libs/vm_monitor/src/runner.rs:97]: It's unclear to me what's wiggling here. Please consider improving the comment on the config variable and this ocmment.

**Comment by problame** [comment]
Looked at the upscale path and left a few comments.

Is there any reason why this still supports the case where file cache sits on tmpfs? I thought we were past that 😮

File: libs/vm_monitor/src/runner.rs (hunk 1 above)
Comment (sharnoff) [reply libs/vm_monitor/src/runner.rs:420]: Oh yikes, I hadn't realized that before. And yeah, it generally isn't worth making the cgroups stuff optional, just a holdover from earlier designs.

Unfortunately, this bad `tokio::select!` is already the current behavior, which was implemented in #5441, so I'm inclined to ignore this for now and address it in a follow-up.
(Broadly there's significant refactoring that I'd like to do here [particularly in pursuit of neondatabase/cloud#6701], but haven't had the chance to get to that yet)

File: libs/vm_monitor/src/runner.rs (hunk 3 above)
Comment (sharnoff) [reply libs/vm_monitor/src/runner.rs]: improved the comment here. config variable will be removed along with support for tmpfs in a follow-up PR

**Comment by sharnoff** [comment]
> I assume theses tests happened?

Only one of them had. The second one (manually testing by artificial memory usage) has now been done, and this PR works as expected. Testing process below:

<details>
<summary>Testing process</summary>

With a checked out autoscaling repo, change the Makefile's `VM_MONITOR_BRANCH` from `main` to `sharnoff/vm-monitor-memory-polling`.

Set up the cluster and prep the test VM's image:

```sh
make kind-setup && make deploy && make pg14-disk-test
```

Deploy the test VM:

```sh
kubectl apply -f vm-deploy.yaml
```

Open multiple windows to stream logs and current VM state. In one, run:

```sh
kubectl get -w neonvm
```

and in another, run:

```sh
kubectl logs -f -n kube-system -l name=autoscaler-agent | jq --unbuffered -r '"\(.ts | strftime("%Y-%m-%d %T")) \(.level) \(.logger): \(.msg) err=\(.error) stack=\(.stack)"'
```

---

Log into the VM with `kubectl exec -it <pod name> -- screen /dev/pts/0`, and run:

```sh
cgexec -g memory:neon-test allocate-loop 800 850
```

The VM should immediately upscale, and then `^C` to kill `allocate-loop` should cause it to downscale shortly after (within a few seconds; but not as quickly as the upscaling).

</details>

> Is there any reason why this still supports the case where file cache sits on tmpfs? I thought we were past that 😮

Yeah; rollout was gradual, via a per-region config flag, so even though the flag has been enabled for every region now (since last Wednesday), the old code has not yet been removed (partly because I wasn't sure that we wouldn't find issues and subsequently need to revert away from on-disk).

General plan in that regard is that removing support for tmpfs-backed file cache _here_ will come after removing the config flag in the control plane, just out of an abundance of caution in case there's any misconfigurations.

File: libs/vm_monitor/src/runner.rs [hunk 4]
```diff
@@ -267,24 +291,18 @@ impl Runner {
...
-                bytes_to_mebibytes(new_cgroup_mem_high),
-                bytes_to_mebibytes(available_memory)
+                "set cgroup memory threshold from {} MiB to {} MiB, of new total {} MiB",
+                bytes_to_mebibytes(cgroup.threshold),
+                bytes_to_mebibytes(new_threshold),
+                bytes_to_mebibytes(usable_system_memory)
             );
+            cgroup.threshold = new_threshold;
             info!("downscale: {message}");
             status.push(message);
```
Comment (problame) [inline libs/vm_monitor/src/runner.rs:307]: Where do we actually reconfigure the cgroup? Do we even still do that?

**Review by problame — COMMENTED** [review]
Took a look at downscale, turned out confused

File: libs/vm_monitor/src/runner.rs (hunk 4 above)
Comment (sharnoff) [reply libs/vm_monitor/src/runner.rs:307]: After this PR, the only interaction we have with the cgroup is to fetch memory statistics. I guess saying "set cgroup memory threshold" is a bit misleading in that sense

**Review by problame — APPROVED** [review]
Synced offline, good to go.

**Review by problame — APPROVED** [review]
Synced offline, good to go.
